pyswarms/backend/handlers.py

```diff
@@ -0,0 +1,163 @@
+"""
+
+import logging
+
+import numpy as np
+
+from ..util.reporter import Reporter
+
+rep = Reporter(logger=logging.getLogger(__name__))
+
+
+class BoundaryHandler:
+    def __init__(self, strategy):
+        """ A BoundaryHandler class
+
+        This class offers a way to handle boundary conditions. It contains
+        methods to avoid having particles outside of the defined boundaries.
+        It repairs the position of particles that would leave the boundares
+        in the next optimization step by using one of the follwing methods:
+
+        * Nearest:
+            Reposition the particle to the nearest bound.
+        * Random:
+            Reposition the particle randomly in between the bounds.
+        * Shrink:
+            Shrink the velocity of the particle such that it lands on the
+            bounds.
+        * Reflective:
+            Mirror the particle position from outside the bounds to inside the
+            bounds.
+        * Intermediate:
+            Reposition the particle to the midpoint between its current
+            position on the bound surpassing axis and the bound itself.
+            This only adjusts the axes that surpass the boundaries.
+        * Resample:
+            Redraw the velocity until the next position is inside the bounds.
+
+        Attributes
+        ----------
+        strategy : str
+            The strategy to be used. The following are available:
+            * "nearest"
+            * "random"
+            * "shrink"
+            * "reflective"
+            * "intermediate"
+            * "resample"
+            For a description of these see above.
+        """
+        self.strategy = strategy
+
+    def __call__(self, position, bounds, *args, **kwargs):
+        """Make class callable
+
+        The BoundaryHandler can be called as a function to use the strategy
+        that is passed at initialization to repair boundary issues. An example
+        for the usage:
+
+        .. code-block :: python
+
+            from pyswarms.backend import operators as op
+            from pyswarms.backend.handlers import BoundaryHandler
+
+            bh = BoundaryHandler(strategy="reflective")
+            ops.compute_position(swarm, bounds, handler=bh)
+
+
+        Parameters
+        ----------
+        position : np.ndarray
+            The swarm position to be handled
+        bounds : tuple of :code:`np.ndarray` or list
+            a tuple of size 2 where the first entry is the minimum bound while
+            the second entry is the maximum bound. Each array must be of shape
+            :code:`(dimensions,)`
+        *args : tuple
+        **kwargs : dict
+
+        Returns
+        -------
+        numpy.ndarray
+            the adjusted positions of the swarm
+        """
+        # Assign new attributes
+        self.position = position
+        self.lower_bound, self.upper_bound = bounds
+        self.__out_of_bounds()
+
+        if self.strategy == "nearest":
```

Comment:
Maybe it's better to just pass `position`, `bounds`, `*args` and`**kwargs` to each of these methods, so that it's easier to test them later on.

Reply:
I see, I'm not quite sure what we're going to do with the args and kwargs, though? We either need the positions or the velocity not much more :confused:

Reply:
As a note, this is also what I'm planning to achieve in the `GeneralOptimizer` (#232). We want to remove all those `if-else`'s since it becomes a handful given different topologies requiring different sets of inputs. Maybe the `compute_*` methods should just maintain the same call as with the base class, and have all others be delegated as a `**kwargs` argument.

Reply:
Thus, if we only have this as our main call:
```python
try:
    new_position = self.strategies[strategy](position, bounds, *args, **kwargs)
except KeyError:
    rep.log("Something") # Logger error 
    raise
else:
    return new_position
```

everything becomes much easier